libcudacxx/include/cuda/__functional/address_stability.h

```diff
@@ -0,0 +1,70 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _CUDA___FUNCTIONAL_ADDRESS_STABILITY_H
+#define _CUDA___FUNCTIONAL_ADDRESS_STABILITY_H
+
+#include <cuda/std/detail/__config>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#include <cuda/std/__type_traits/integral_constant.h>
+#include <cuda/std/__type_traits/void_t.h>
+#include <cuda/std/__utility/move.h>
+
+_LIBCUDACXX_BEGIN_NAMESPACE_CUDA
+
+// need a separate implementation trait because we SFINAE with a type parameter before the variadic pack
+template <typename F, typename SFINAE, typename... Args>
+struct __allows_copied_arguments_impl : _CUDA_VSTD::false_type
+{};
+
+template <typename F, typename... Args>
+struct __allows_copied_arguments_impl<F, _CUDA_VSTD::void_t<decltype(F::allows_copied_arguments)>, Args...>
+{
+  static constexpr bool value = F::allows_copied_arguments;
+};
+
+//! Trait telling whether a function object relies on the memory address of its arguments when called with the given set
+//! of types. The nested value is true when the addresses of the arguments do not matter and arguments can be provided
+//! from arbitrary copies of the respective sources. Can be specialized for custom function objects and parameter types.
+template <typename F, typename... Args>
+struct allows_copied_arguments : __allows_copied_arguments_impl<F, void, Args...>
+{};
```

Comment:
In what capacity are the variadic function arguments ever used? Could we just drop them?

Reply:
Two cases come to my mind:

1. You can construct a pathological case where you have a function object with an overload set where one overload permits copied arguments and the other doesn't:
```c++
struct overloaded
{
  // permits copies
  _CCCL_HOST_DEVICE auto operator()(int a, int b) const -> int
  {
    return a + b;
  }

  // does not permit copies, requires original address
  _CCCL_HOST_DEVICE auto operator()(const float& a, const float& b) const -> float
  {
    return (float)(uintptr_t)&a + b;
  }
};
```
The variadic version of `allows_copied_arguments` would allow to distinguish the overloads and only proclaim the top one.

2. Because I can only see the signatue, e.g. `operator()(const float& a, const float& b)`, if the input arguments are actually not `float`, but `int`, a temporary is produced anyway to which the `const float&` binds, indicating that the `operator()` cannot rely on the address of it's argument without being buggy. So I not only need the function signature, I also need the argument types.

However, both cases are rather uncommon. In the interest of better usability I could also imagine dropping the whole idea and allowing only:
```c++
template <typename F>
struct allows_copied_arguments;
```
And if you have function object with a pathological overload set, you are just out of luck. You can always create a new function object and call your pathological one inside as a workaround.